pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -1143,15 +1142,24 @@ func (cb *ClusterBuilder) buildUpstreamClusterTLSContext(opts *buildClusterOpts,
 	return tlsContext, nil
 }
 
-func (cb *ClusterBuilder) setAutoSni(mc *MutableCluster, tls *networking.ClientTLSSettings) {
-	if mc != nil && features.EnableAutoSni && len(tls.Sni) == 0 {
-		if mc.httpProtocolOptions == nil {
-			mc.httpProtocolOptions = &http.HttpProtocolOptions{}
-		}
-		if mc.httpProtocolOptions.UpstreamHttpProtocolOptions == nil {
-			mc.httpProtocolOptions.UpstreamHttpProtocolOptions = &core.UpstreamHttpProtocolOptions{}
+// Set auto_sni if sni field is not explicitly set and if EnableAutoSni feature flag is enabled.
+// Set auto_san_validation if VerifyCertAtClient feature flag is enabled
+func (cb *ClusterBuilder) setAutoSniAndAutoSanValidation(mc *MutableCluster, tls *networking.ClientTLSSettings) {
+	if mc != nil && features.EnableAutoSni {
+		if len(tls.Sni) == 0 || (features.VerifyCertAtClient && len(tls.SubjectAltNames) == 0) {
```

Comment:
should not we return with some warning message if EnableAutoSni is true and `len(tls.Sni) != 0`?

Reply:
DR allows setting the sni explicitly, in that case we donot want to use auto_sni, and that is an allowed usecase. So I am not sure if we need a warning message in that case.